### PR TITLE
introduce `SubPcGroup`

### PR DESCRIPTION
**Pull request by ThomasBreuer**
This is a first attempt to implement what was sketched in #2672.
I am not sure that we really want this:
- Admitting different types for a group and a subgroup was apparently not allowed up to now, several signatures of functions have to be "relaxed" in order to support different types.
- Functions returning arrays of subgroups (typically series such as `derived_series`) will now create the full group (if it occurs) with another type; one has to be aware of different types for the same mathematical object.

If we want to proceed like this then the next step is as follows.
- Fix functions like `intersection` which expect arguments of the same type up to now.
- Adjust the doctests.
- Adjust/extend the documentation.
- Add tests.
- Do the same for `FpGroup`/`SubFpGroup`.

### CODE REVIEW

File: src/Groups/GAPGroups.jl
Comment (lgoettgens) [inline src/Groups/GAPGroups.jl]:
```suggestion
core(G::GAPGroup, H::GAPGroup) = _as_subgroup(G, GAPWrap.Core(G.X, H.X))
```
Your change makes it actually a lot easier

**Comment by fingolfin** [comment]
> This is a first attempt to implement what was sketched in #2672. I am not sure that we really want this:
> 
> * Admitting different types for a group and a subgroup was apparently not allowed up to now, several signatures of functions have to be "relaxed" in order to support different types.

That seems fine to me.

> * Functions returning arrays of subgroups (typically series such as `derived_series`) will now create the full group (if it occurs) with another type; one has to be aware of different types for the same mathematical object.

I also think this is fine, and in fact just reflects the reality in GAP, which tries to paper of it but doesn't really succeed. So I actually think it is *better* that we make this explicit. This way at least the user can immediately tell if a given group is the "full" group or not, and we have a chance of explaining the situation in the docstring for `SubPcGroup`.

> 
> If we want to proceed like this then the next step is as follows.
> 
> * Fix functions like `intersection` which expect arguments of the same type up to now.
> * Adjust the doctests.
> * Adjust/extend the documentation.
> * Add tests.
> * Do the same for `FpGroup`/`SubFpGroup`.

File: src/Groups/GAPGroups.jl
Comment (fingolfin) [inline src/Groups/GAPGroups.jl]: Maybe like this?
```suggestion
  T = typeof(G) == PcGroup ? "Pc group" : "Subgroup of pc group"
```

File: src/Groups/GAPGroups.jl
Comment (fingolfin) [reply src/Groups/GAPGroups.jl]: It makes the signatures easier (with Lars' extra change). Of course we loose a bit of "validation", i.e. previously Julia prevented some situations from occurring (e.g. you could not mix a permutation group with a matrix group), and now we have to catch that ourselves. But that's not so bad in the end, we already have to catch some of these "bad" situations, e.g. when both groups are matrix groups but defined over different rings (perhaps in different characteristic)

File: src/Groups/group_constructors.jl
Comment (fingolfin) [inline src/Groups/group_constructors.jl:121]: This creates a new `PcGroup`, which would be the "whole" group, so should we really allow `SubPcGroup` here?

File: src/Groups/homomorphisms.jl
Comment (fingolfin) [inline src/Groups/homomorphisms.jl:594]: This, by the way, is something that I needed a lot in the past, and was always annoyingly complicated in GAP. So this is a nice bonus I did not even consider (at least I don't recall considering it :-) )

File: src/Groups/homomorphisms.jl
Comment (fingolfin) [inline src/Groups/homomorphisms.jl]: What about `pcp` groups, they don't use `FamilyPcgs` and probably needs separate code?

File: src/Groups/homomorphisms.jl
Comment (fingolfin) [inline src/Groups/homomorphisms.jl]: Do we really want/need this? Maybe `SubPcGroup` on a `PcGroup` could be useful, in case one needs the "full group as a subgroup", but other than that?

File: src/Groups/types.jl
Comment (fingolfin) [inline src/Groups/types.jl:280]: To me the most annoying part of this PR actually seems to be `SubPcGroupElem`; we probably would want to allow multiplying a `SubPcGroupElem` by a `PcGroupElem` (and vice-versa), and possibly in various other contexts one may have to / wish to support both?

Perhaps a suitable `promote_rule` can help a bit?

File: src/Groups/homomorphisms.jl
Comment (ThomasBreuer) [reply src/Groups/homomorphisms.jl:594]: I think GAP is just missing something like `IsomorphismFullPcGroup`.

File: src/Groups/sub.jl
Comment (fingolfin) [inline src/Groups/sub.jl]: We could add a helper
```julia
_is_subgroup_type(::Type{T1}, ::Type{T2}) where {T1, T2} = false
_is_subgroup_type(::Type{T}, ::Type{T}) where {T <: GAPGroup} = true
_is_subgroup_type(::Type{PcGroup}, ::Type{SubPcGroup}) = true
```
and then add checks like
```julia
  @req _is_subgroup_type(T1, T2) "argument types do not fit"
```

**Comment by fingolfin** [comment]
Just to document it somewhere in writing: We discussed this some time ago and the verdict was that yes, we want to proceed in this direction; while there are downsides, many of them actually already exists, but just are hidden; and there are also upsides. So...

File: src/Groups/sub.jl
Comment (ThomasBreuer) [reply src/Groups/sub.jl]: I thought about introducing something like `_compatibility_check(G1::T1, G2::T2)` which throws an exception if the arguments do not fit together, and otherwise has no effect. Note that checking the types alone is not sufficient, for example in the case of two f.p. subgroups or two pc subgroups which belong to different full groups. If we do not check this compatibility before delegating to GAP then we get just a cryptic error message from GAP.

**Comment by fingolfin** [comment]
@ThomasBreuer I hope this is still on your radar? :-)

**Comment by ThomasBreuer** [comment]
The failure of one of the book examples shows a conceptual problem. (Likely this example can be turned back to the old behaviour by an improvement on the GAP side, but this does not solve the general problem.)

The function `quo` for two groups states that the type of the return value depends on the situation: Anyhow a new group must be created, and GAP decides about a suitable representation; if the result is known to be a finite solvable group then a pc group is returned, otherwise a permutation group or a fp group are possible.

The GAP function `NaturalHomomorphismByNormalSubgroup` is allowed to return a mapping that is not (known to be) surjective. If this happens and if the `Range` of the mapping is a pc group then the quotient on the Oscar side can be only a `SubPcGroup`. (At some extra cost, we can check the surjectivity, and create a `PcGroup` in Oscar if yes.)

It looks as if we have to do something more specific on the GAP side if we want the results of `quo` to be either `PcGroup` or `FPGroup` or `PermGroup`.

(@fieker @fingolfin ?)

**Comment by ThomasBreuer** [comment]
The failure in `test (1.10, book, ubuntu-latest)` is due to an intended change of the output shown: The group in question has now a different type.

**Comment by benlorenz** [comment]
> The failure in `test (1.10, book, ubuntu-latest)` is due to an intended change of the output shown: The group in question has now a different type.

Then you can adjust the output in (`test/book/cornerstones/groups/actions.jlcon`) to the new printing. Only the input must stay identical for 1.1 (i.e. current master).

File: src/Groups/group_constructors.jl
Comment (fingolfin) [reply src/Groups/group_constructors.jl:121]: @ThomasBreuer did you see my question from January there? I don't terribly mind having this (it may make some generic code easier to write), but I wonder: does `cyclic_group(SubPcGroup, 5)` really work correctly, and return something of type `SubPcGroup`? We don't have a test for that, do we?

File: src/Groups/homomorphisms.jl
Comment (fingolfin) [inline src/Groups/homomorphisms.jl]:
```suggestion
         f = GAP.Globals.IsomorphismPcpGroup(GapObj(G))::GapObj
```
possibly also similar (i.e. `FOO.X` to `GapObj(FOO)`) elsewhere?

File: src/Groups/homomorphisms.jl
Comment (fingolfin) [inline src/Groups/homomorphisms.jl]: Huh, does this line work as intended? There is no test case covering this code according to Codecov, and I have some doubts: won't this create a GAP list of OSCAR elements?

Perhaps like this:
```suggestion
         Ggens = GapObj(gens(G); recursive=true)
```

File: src/Groups/homomorphisms.jl
Comment (fingolfin) [inline src/Groups/homomorphisms.jl:709]: I guess we can drop the `GAP.` from `GAP.GapObj` here and elsewhere? Then again we still use it in a bunch of places, so we could change that in a later follow-up PR.
```suggestion
         GapG = GAP.Globals.AbelianPcpGroup(length(exponents), GapObj(exponents, recursive=true))
```

File: src/Groups/homomorphisms.jl
Comment (fingolfin) [inline src/Groups/homomorphisms.jl]:
```suggestion
```

File: src/Groups/matrices/MatGrp.jl
Comment (fingolfin) [inline src/Groups/matrices/MatGrp.jl:438]: If we don't need it, remove it? (Though perhaps in a separate PR?)

Though then again, now I wonder if the general one shouldn't be re-implemented by calling `GAP.Globals.Comm`? And then it might make sense to keep the above method here... Or even change it to, say, `inv(y*x)*(x*y)` (one less inversion, which is arguably the most expensive bit here?)

File: src/Groups/sub.jl
Comment (fingolfin) [inline src/Groups/sub.jl:194]: Hmmm... this is ambiguous, isn't it? The `4` refers to the order of the subgroup, but a reader might reasonably interpret this as saying the parent group has order 4... :-/.

(How) can we make this clearer? Perhaps We should call it a `Sub-pc group` (and explain that somewhere)? 

Or use a different phrasing: `Subgroup of pc group, has order 4` -- alas that would be different from how this kind of info is printed for almost anything else in OSCAR, wouldn't it?

Perhaps @thofma or @fieker have some brilliant insight on this.

File: src/Groups/sub.jl
Comment (fingolfin) [inline src/Groups/sub.jl]:
```suggestion
  if !(GAPWrap.IsPcGroup(cod) && GAPWrap.IsSurjective(mp))
```

File: src/Groups/sub.jl
Comment (fingolfin) [inline src/Groups/sub.jl]:
```suggestion
  cod = GAPWrap.Range(mp)
```

File: src/Groups/sub.jl
Comment (fingolfin) [inline src/Groups/sub.jl]: Maybe "GAP wrap" `ImagesSource`, too?

File: src/Groups/types.jl
Comment (fingolfin) [inline src/Groups/types.jl:373]: I agree: `_oscar_subgroup` would be better

File: src/Groups/types.jl
Comment (fingolfin) [inline src/Groups/types.jl]: This is not used anywhere, is it? Can we just remove it?

File: test/Groups/homomorphisms.jl
Comment (fingolfin) [inline test/Groups/homomorphisms.jl]: What is the issue here? Can we perhaps at least record that in the comment?
I.e. do these run into errors? Or return wrong results? Or something else? If there is a relevant issue, link to it here?

File: test/Groups/libraries.jl
Comment (fingolfin) [inline test/Groups/libraries.jl:144]: Is there a point in keep the "old" copies of the test around, here and below? Otherwise just delete them?

File: test/Groups/quotients.jl
Comment (fingolfin) [inline test/Groups/quotients.jl:58]: Why not
```suggestion
   for T in [ PermGroup, PcGroup, SubPcGroup ]
```

**Review by fingolfin — COMMENTED** [review]
Some remarks. Most are optional (better to get this merged sooner and then polish afterwards), but there are one or two were I wonder if the code is correct as it is...?

File: src/Groups/homomorphisms.jl
Comment (ThomasBreuer) [reply src/Groups/homomorphisms.jl]: `gens(G)` consists of OSCAR elements, the array of the `x.X` consists of GAP elements.
I am going to add tests.
(And yes, the code you propose is better.)

File: src/Groups/homomorphisms.jl
Comment (ThomasBreuer) [reply src/Groups/homomorphisms.jl]: O.k., I change all those `.X` introduced in this pull request; if I change also others then there will likely be conflicts.

File: src/Groups/homomorphisms.jl
Comment (ThomasBreuer) [reply src/Groups/homomorphisms.jl:709]: Yes, we can write just `GapObj` because meanwhile it is an exported variable.

File: src/Groups/matrices/MatGrp.jl
Comment (ThomasBreuer) [reply src/Groups/matrices/MatGrp.jl:438]: Let us leave this to another pull request.

File: src/Groups/types.jl
Comment (ThomasBreuer) [reply src/Groups/types.jl]: Yes.

File: test/Groups/quotients.jl
Comment (ThomasBreuer) [reply test/Groups/quotients.jl:58]: The point is that currently `abelian_group(PcGroup, [2, 3, 4])` is not supported.
`abelian_group` promises a result whose generators correspond to the given orders,
and the type `PcGroup` implies that the generators are a polycyclic generating sequence w.r.t. the underlying presentation.
With GAP's `IsPcGroup`, we cannot get this; and `IsPcpGroup` cannot be used because of some bugs.

(I will add a comment.)

File: test/Groups/libraries.jl
Comment (ThomasBreuer) [reply test/Groups/libraries.jl:144]: I am still confident that eventually we can support `abelian_group(PcGroup, [2, 4])`.
I will add a comment.

File: test/Groups/homomorphisms.jl
Comment (ThomasBreuer) [reply test/Groups/homomorphisms.jl]: I will add a comment.

File: src/Groups/types.jl
Comment (ThomasBreuer) [reply src/Groups/types.jl:373]: O.k., I will rename the two-argument version.

**Comment by fingolfin** [comment]
Unfortunately this has conflicts again now :-(

It would be great if we could merge this tomorrow (before or after triage, whenever you have time to resolve it) so we can move on :-)

File: src/Groups/sub.jl
Comment (fingolfin) [reply src/Groups/sub.jl:194]: To be clear: we can address this in a follow-up PR (but then let's make an issue about this after the PR is merged so so we don't forget)

**Comment by ThomasBreuer** [comment]
> Unfortunately this has conflicts again now :-(

I find it rather surprising that there were not more conflicts.

**Comment by fingolfin** [comment]
@ThomasBreuer there is one unresolved review comment of mine from quite some ago where I ask about `sub_pc_group` -- can you please reply to that?

File: src/Groups/sub.jl
Comment (ThomasBreuer) [reply src/Groups/sub.jl:194]: I have opened issue  #3732.

File: src/Groups/group_constructors.jl
Comment (ThomasBreuer) [reply src/Groups/group_constructors.jl:121]: Yes, `cyclic_group(SubPcGroup, 5)` works.
(I had intended to add a test for this, but somehow this got lost.)

**Review by fingolfin — APPROVED** [review]

**Pull request merged**
